### PR TITLE
Allow old certs to be cross-signed

### PR DESCRIPTION
**Pull request by cipherboy**
In Vault 1.11, we introduced cross-signing support, but the earlier SKID
field change in Vault 1.10 causes problems: notably, certs created on
older versions of Vault (<=1.9) or outside of Vault (with a different
SKID method) cannot be cross-signed and validated in OpenSSL.

In particular, OpenSSL appears to be unique in requiring a SKID/AKID
match for chain building. If AKID and SKID are present on an otherwise
valid client/parent cert pair and the values are different, OpenSSL will
not build a valid path over those two, whereas most other chain
validation implementations will.

Regardless, to have proper cross-signing support, we really aught to
support copying an SKID. This adds such support to the sign-intermediate
endpoint. Support for the /issue endpoint is not added, as cross-signing
leaf certs isn't generally useful and can accept random SKIDs.

Resolves: #16461

`Signed-off-by: Alexander Scheel <alex.scheel@hashicorp.com>`

### CODE REVIEW

**Comment by cipherboy** [comment]
@dmitriy-moiseev -- do you want to test this to make sure this satisfies your use case? If you go to the [GH test build -> summary page](https://github.com/hashicorp/vault/actions/runs/2756614713) you can fetch a pre-built scratch binary if it is of interest.

File: builtin/logical/pki/path_sign_issuers.go [hunk 1]
```diff
@@ -67,6 +67,24 @@ SHA-2-512. Defaults to 0 to automatically detect based on key length
 		},
 	}
 
+	fields["skid"] = &framework.FieldSchema{
+		Type:    framework.TypeString,
+		Default: 0,
```
Comment (stevendpclark) [inline builtin/logical/pki/path_sign_issuers.go:72]: Should this not be a blank string instead of a 0?

Comment (cipherboy) [reply builtin/logical/pki/path_sign_issuers.go:72]: Ah, no, you're correct.

File: builtin/logical/pki/cert_util.go [hunk 2]
```diff
@@ -1340,6 +1341,24 @@ func generateCreationBundle(b *backend, data *inputBundle, caSign *certutil.CAIn
...
+	var skid []byte
+	{
+		if rawSKIDValue, ok := data.apiData.GetOk("skid"); ok {
+			var err error
+			skidValue := rawSKIDValue.(string)
+
+			// Handle removing common separators to make copy/paste from tool
+			// output easier.
+			skidNoColonsSeparators := strings.ReplaceAll(skidValue, ":", "")
+			skidNoSeparators := strings.ReplaceAll(skidNoColonsSeparators, "-", "")
```
Comment (kitography) [inline builtin/logical/pki/cert_util.go:1354]: Could we add space here? (That's what's used in browser decoding)

File: builtin/logical/pki/chain_test.go [hunk 3]
```diff
@@ -151,6 +153,14 @@ func (c CBGenerateIntermediate) Run(t testing.TB, b *backend, s logical.Storage,
 	if len(c.CommonName) > 0 {
 		data["common_name"] = c.CommonName
 	}
+	if len(c.SKID) > 0 {
+		// Copy the SKID from an existing, already-issued cert.
+		otherPEM := knownCerts[c.SKID]
+		otherCert := ToCertificate(t, otherPEM)
+
+		data["skid"] = hex.EncodeToString(otherCert.SubjectKeyId)
```
Comment (kitography) [inline builtin/logical/pki/chain_test.go:161]: Realized only here that we can't pass in an existing cert to cross-sign.  We might want to add that?

Nice tests though

**Review by kitography — COMMENTED** [review]
Looks pretty good!

File: builtin/logical/pki/chain_test.go (hunk 3 above)
Comment (cipherboy) [reply builtin/logical/pki/chain_test.go:161]: I'm mostly against it. It is definitely a lot of work (in the existing signing framework via certutil) to add deserialization of existing certificates back into the input parameters format... 

It definitely requires a bit of careful though and structuring...

File: builtin/logical/pki/cert_util.go (hunk 2 above)
Comment (cipherboy) [reply builtin/logical/pki/cert_util.go:1354]: Ah interesting, Chrome uses a space but Firefox/OpenSSL use colons. Sure. :-)

File: builtin/logical/pki/chain_test.go (hunk 3 above)
Comment (cipherboy) [reply builtin/logical/pki/chain_test.go:161]: (I have gone ahead and added a separate test to verify the copy, and I did so here as well).

**Comment by cipherboy** [comment]
@kitography @stevendpclark -- updated!

File: builtin/logical/pki/cert_util.go [hunk 4]
```diff
@@ -1340,6 +1341,25 @@ func generateCreationBundle(b *backend, data *inputBundle, caSign *certutil.CAIn
...
+	{
+		if rawSKIDValue, ok := data.apiData.GetOk("skid"); ok {
+			var err error
+			skidValue := rawSKIDValue.(string)
+
+			// Handle removing common separators to make copy/paste from tool
+			// output easier.
+			skidNoColonsSeparators := strings.ReplaceAll(skidValue, ":", "")
+			skidNoSpaceSeparators := strings.ReplaceAll(skidNoColonsSeparators, " ", "")
+			skidNoSeparators := strings.ReplaceAll(skidNoSpaceSeparators, "-", "")
```
Comment (stevendpclark) [inline builtin/logical/pki/cert_util.go:1355]: Would this be a little easier to extend/mess up by accidentally copy/pasting the wrong line and undoing an existing separator?

```go
skidNoSeparators := skidValue
for _, separator := range []string{":", "-", " "} {
    skidNoSeparators = strings.ReplaceAll(skidNoSeparators, separator, "")
}
```

**Review by stevendpclark — APPROVED** [review]
Looks good to me, one minor nit/suggestion but I don't feel strongly about it.

**Pull request merged**